tcms/signals.py

```diff
@@ -209,19 +209,33 @@ def handle_comments_pre_delete(sender, **kwargs):
 def handle_emails_post_bug_save(sender, instance, created=False, **kwargs):
     """
     Send email updates to assignee after they've been
-    assigned a bug on bug creation.
+    assigned a bug on bug creation or there is a new comment to the bug.
     """
-    if not created or instance.assignee is None:
+    from tcms.core.helpers.comments import get_comments
+
+    comments = get_comments(instance)
+
+    if instance.assignee is None or comments.count() < 1:
+        return
+
+    if (not instance.status and comments.last().comment == _("*bug closed*")) or (
+        instance.status and comments.last().comment == _("*bug reopened*")
+    ):
```

Comment:
Remove top 3 lines. You still want to send notifications even when a bug is closed or reopened.

Reply:
>  one during add_comment 

That was your older implementation which is now gone. This shouldn't be happening b/c as far as I can see there's nothing to handle add_comment and send emails. Maybe you have some old code (could also be .pyc) lurking around?

Reply:
Correction to previous comment. I've misread this. I thought you meant your previous custom signal definition, not the `add_comment()` helper function. Still even inside that function I don't see email functionality and also I'm not seeing a `post_save` handler for the `Comment` model. 

How do you know that there are 2 emails and how do you know where they come from ?

Reply:
First `post_save` is triggered from here and second is from `bug.save()`
https://github.com/kiwitcms/Kiwi/blob/bd7aa02027dfc127ace494cd5356d9147f99cb45/tcms/core/helpers/comments.py#L54

Resulting two e-mails, comment and the state of the bug mismatches in the first e-mail triggered by add_comment:

> ---------- MESSAGE FOLLOWS ----------
b'Content-Type: text/plain; charset="utf-8"'
b'MIME-Version: 1.0'
b'Content-Transfer-Encoding: 7bit'
b'Subject: [Kiwi-TCMS] Bug #8 - is none check'
b'From: kiwi@example.com'
b'To: emailaddress1@gmail.com, emailaddress2@gmail.com'
b'Date: Sun, 12 Dec 2021 12:04:35 -0000'
b'Message-ID: <163931067580.4328.11001300769411000928@pcHostname>'
b'X-Peer: 127.0.0.1'
b''
b''
b'Bug 8 has been updated.'
b''
b'Link: http://127.0.0.1:8000/bugs/8/'
b''
b'Summary: is none check'
b'Created at: Dec. 1, 2021, 6:49 p.m.'
b'Reporter: cmb'
b'Assignee: mete'
b'Product: Sample Product'
b'Version: V1.0'
b'Build: INC1'
b'Last comment: '
b'*bug reopened*'
------------ END MESSAGE ------------
---------- MESSAGE FOLLOWS ----------
b'Content-Type: text/plain; charset="utf-8"'
b'MIME-Version: 1.0'
b'Content-Transfer-Encoding: 7bit'
b'Subject: [Kiwi-TCMS] Bug #8 - is none check'
b'From: kiwi@example.com'
b'To: emailaddress1@gmail.com, emailaddress2@gmail.com'
b'Date: Sun, 12 Dec 2021 12:04:45 -0000'
b'Message-ID: <163931068590.4328.6952816735701083535@pcHostname>'
b'X-Peer: 127.0.0.1'
b''
b''
b'Bug 8 has been updated.'
b''
b'Link: http://127.0.0.1:8000/bugs/8/'
b''
b'Summary: is none check'
b'Created at: Dec. 1, 2021, 6:49 p.m.'
b'Reporter: cmb'
b'Assignee: mete'
b'Product: Sample Product'
b'Version: V1.0'
b'Build: INC1'
b'Last comment: '
b'*bug reopened*'
------------ END MESSAGE ------------

Which is also visible from the call stack:

![add_comment_post_save](https://user-images.githubusercontent.com/8937774/145711640-f08cebdb-a182-43cd-a4d6-3a2effc1015d.png)

Any hint to ignore duplicate e-mails would be appreciated.

Reply:
`d15304afbd0c4f851450b1ff716efb2179fb78e7` is the root cause of the double emails. That is an important commit which deals with inline attachments from the rich text editor and we can't change the behavior without breaking everything else. 

This has existed for a while and I suspect the reason why we haven't seen duplicate emails before is a coincidence from the previous if conditions and how the email handling code was implemented. Now that we're changing the behavior the problem is being exposed. 

IMO the new implementation in this PR should be the correct one, e.g. send all notifications on bug changes regardless of the actual changes. The duplicate emails are a side issue from all of the current changes.

That said I don't think there is a quick and easy way to resolve the issue of duplicate emails. My vote goes for merging this PR and living with the duplicate emails as a bug for a while.